mm2src/coins/eth.rs

```diff
@@ -2670,150 +2683,158 @@ impl EthCoin {
+                return MmError::err(ValidatePaymentError::UnexpectedPaymentState(format!(
+                    "Payment state is not PAYMENT_STATE_SENT, got {}",
+                    status
+                )));
             }
 
-            let tx_from_rpc = try_s!(
-                selfi
-                    .web3
-                    .eth()
-                    .transaction(TransactionId::Hash(tx.hash))
-                    .compat()
-                    .await
-            );
+            let tx_from_rpc = selfi
+                .web3
+                .eth()
+                .transaction(TransactionId::Hash(tx.hash))
+                .compat()
+                .await?;
             let tx_from_rpc = match tx_from_rpc {
                 Some(t) => t,
-                None => return ERR!("Didn't find provided tx {:?} on ETH node", tx),
+                None => {
+                    return MmError::err(ValidatePaymentError::MissingTx(format!(
+                        "Didn't find provided tx {:?} on ETH node",
+                        tx
+                    )))
+                },
             };
 
             if tx_from_rpc.from != sender {
-                return ERR!(
+                return MmError::err(ValidatePaymentError::WrongSenderAddress(format!(
                     "Payment tx {:?} was sent from wrong address, expected {:?}",
-                    tx_from_rpc,
-                    sender
-                );
+                    tx_from_rpc, sender
+                )));
             }
 
             match &selfi.coin_type {
                 EthCoinType::Eth => {
                     if tx_from_rpc.to != Some(expected_swap_contract_address) {
-                        return ERR!(
+                        return MmError::err(ValidatePaymentError::WrongReceiverAddress(format!(
                             "Payment tx {:?} was sent to wrong address, expected {:?}",
-                            tx_from_rpc,
-                            expected_swap_contract_address
-                        );
+                            tx_from_rpc, expected_swap_contract_address
+                        )));
                     }
 
                     if tx_from_rpc.value != expected_value {
-                        return ERR!(
+                        return MmError::err(ValidatePaymentError::InvalidPaymentTxData(format!(
                             "Payment tx {:?} value is invalid, expected {:?}",
-                            tx_from_rpc,
-                            expected_value
-                        );
+                            tx_from_rpc, expected_value
+                        )));
                     }
 
-                    let function = try_s!(SWAP_CONTRACT.function("ethPayment"));
-                    let decoded = try_s!(function.decode_input(&tx_from_rpc.input.0));
+                    let function = SWAP_CONTRACT.function("ethPayment")?;
+                    let decoded = function.decode_input(&tx_from_rpc.input.0)?;
```

Comment:
As I can see from this implementation:
`fn from(err: ethabi::Error) -> Self { Self::AbiError(err.to_string()) }`, `validate_payment` returns `ValidatePaymentError::AbiError` here.
I think we can simply return `InvalidPaymentTxData`.